crypto/fift/HashMap.h

```diff
@@ -0,0 +1,306 @@
+      : key_(std::move(key)), value_(std::move(value)), left_(std::move(left)), right_(std::move(right)), y_(y) {
+  }
+  Hashmap(const Hashmap& other) = default;
+  Hashmap(Hashmap&& other) = default;
+  virtual ~Hashmap() {
+  }
+  Hashmap* make_copy() const override {
+    return new Hashmap(*this);
+  }
+  const DictKey& key() const& {
+    return key_;
+  }
+  DictKey key() && {
+    return std::move(key_);
+  }
+  const vm::StackEntry& value() const& {
+    return value_;
+  }
+  vm::StackEntry value() && {
+    return std::move(value_);
+  }
+  Ref<Hashmap> left() const {
+    return left_;
+  }
+  Ref<Hashmap> right() const {
+    return right_;
+  }
+  Ref<Hashmap> lr(bool branch) const {
+    return branch ? right_ : left_;
+  }
+  Ref<Hashmap> rl(bool branch) const {
+    return branch ? left_ : right_;
+  }
+  static Ref<Hashmap> lookup_key(Ref<Hashmap> root, const DictKey& key);
+  template <typename... Args>
+  static Ref<Hashmap> lookup(Ref<Hashmap> root, Args&&... args) {
+    return lookup_key(std::move(root), DictKey{std::forward<Args>(args)...});
+  }
+  static vm::StackEntry get_key(Ref<Hashmap> root, const DictKey& key);
+  template <typename... Args>
+  static vm::StackEntry get(Ref<Hashmap> root, Args&&... args) {
+    return get_key(std::move(root), DictKey{std::forward<Args>(args)...});
+  }
+  static Ref<Hashmap> remove_key(Ref<Hashmap> root, const DictKey& key);
+  template <typename... Args>
+  static Ref<Hashmap> remove(Ref<Hashmap> root, Args&&... args) {
+    return remove_key(std::move(root), DictKey{std::forward<Args>(args)...});
+  }
+  static std::pair<Ref<Hashmap>, vm::StackEntry> get_remove_key(Ref<Hashmap> root, const DictKey& key);
+  template <typename... Args>
+  static std::pair<Ref<Hashmap>, vm::StackEntry> get_remove(Ref<Hashmap> root, Args&&... args) {
+    return get_remove_key(std::move(root), DictKey{std::forward<Args>(args)...});
+  }
+  static Ref<Hashmap> set(Ref<Hashmap> root, const DictKey& key, vm::StackEntry value);
+  static bool replace(Ref<Hashmap>& root, const DictKey& key, vm::StackEntry value);
+  static std::pair<Ref<Hashmap>, Ref<Hashmap>> split(Ref<Hashmap> root, const DictKey& key, bool eq_left = false);
+  static Ref<Hashmap> empty() {
+    return {};
+  }
+
+ private:
+  static Ref<Hashmap> merge(Ref<Hashmap> a, Ref<Hashmap> b);
+  static const Hashmap* lookup_key_aux(const Hashmap* root, const DictKey& key);
+  Ref<Hashmap> get_remove_internal(const DictKey& key, vm::StackEntry& val) const;
+  Ref<Hashmap> replace_internal(const DictKey& key, const vm::StackEntry& value, bool& found) const;
+  static void insert(Ref<Hashmap>& root, const DictKey& key, vm::StackEntry value, long long y);
+  static long long new_y();
+};
+
+struct HashmapIdx {
+  Ref<Hashmap>& root_;
+  DictKey idx_;
+  template <typename... Args>
+  HashmapIdx(Ref<Hashmap> root, Args&&... args) : root_(root), idx_(std::forward<Args>(args)...) {
```

Comment:
Pass `Ref<Hashmap> &root` instead of `Ref<Hashmap> root`.
If `root` is passed by value, `root_` will refer to parameter `root` which is destroyed after the constructor.